src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs

```diff
@@ -0,0 +1,3144 @@
+            ElseIf b = 3 Then
+                i.Last()
+                [|j|].LastOrDefault()
+            Else
+                i.LongCount()
+                [|j|].Max()
+            End If
+
+            i.Min()
+            [|j|].Single()
+        End Sub
+    End Class
+End Namespace
+";
+            await VerifyVB.VerifyAnalyzerAsync(vbCode);
+        }
+
+        [Fact]
+        public async Task TestInvocationAfterIfbranch()
+        {
+            var csharpCode = @"
+using System.Collections.Generic;
+using System.Linq;
+
+public class Bar
+{
+    public void Sub(IEnumerable<int> h, int x)
+    {
+        IEnumerable<int> i = Enumerable.Range(1, 10);
+        if (x == 1)
+        {
+            [|i|].Except([|h|]).ToList();
+            [|i|].Intersect([|h|]).ToList();
+        }
+
+        [|i|].Max();
+    }
+}";
+            await VerifyCS.VerifyAnalyzerAsync(csharpCode);
+
+            var vbCode = @"
+Imports System.Collections.Generic
+Imports System.Linq
+
+Namespace Ns
+    Public Class Hoo
+        Public Sub Goo(h As IEnumerable(Of Integer), x As Integer)
+            Dim i = Enumerable.Range(1, 10)
+            If x = 1 Then
+                [|i|].Except([|h|]).ToList()
+                [|i|].Intersect([|h|]).ToList()
+            End If
+
+            [|i|].Max()
+        End Sub
+    End Class
+End Namespace
+";
+            await VerifyVB.VerifyAnalyzerAsync(vbCode);
+        }
+
+        [Fact]
+        public async Task TestInvocationBeforeIfbranch()
+        {
+            var csharpCode = @"
+using System.Collections.Generic;
+using System.Linq;
+
+public class Bar
+{
+    public void Sub(IEnumerable<int> h, int x)
+    {
+        IEnumerable<int> i = Enumerable.Range(1, 10);
+        [|i|].Max();
+        if (x == 1)
+        {
+            [|i|].Except([|h|]).ToList();
+            [|i|].Intersect([|h|]).ToList();
+        }
+
+        [|i|].Max();
+    }
+}";
+            await VerifyCS.VerifyAnalyzerAsync(csharpCode);
+
+            var vbCode = @"
+Imports System.Collections.Generic
+Imports System.Linq
+
+Namespace Ns
+    Public Class Hoo
+        Public Sub Goo(h As IEnumerable(Of Integer), x As Integer)
+            Dim i = Enumerable.Range(1, 10)
+            [|i|].Max()
+            If x = 1 Then
+                [|i|].Except([|h|]).ToList()
+                [|i|].Intersect([|h|]).ToList()
+            End If
+
+            [|i|].Max()
+        End Sub
+    End Class
+End Namespace
+";
+            await VerifyVB.VerifyAnalyzerAsync(vbCode);
+        }
+
+        [Fact]
+        public async Task TestInvocationAfterIfbranchForArray()
+        {
+            var csharpCode = @"
+using System.Collections.Generic;
+using System.Linq;
+
+public class Bar
+{
+    public void Sub(IEnumerable<int> h, int x)
+    {
+        var i = Enumerable.Range(1, 10).ToArray();
+        if (x == 1)
+        {
+            i.Except([|h|]).ToList();
+            i.Intersect([|h|]).ToList();
```

Comment:
For me this 2 calls will be difficult to implement within a one enumeration, i think its better to not flag such scenario (whenever possible)

Reply:
Emm, interesting, so maybe for methods like `except` and 'intersect' should be treated differently. Let me see if I can find a full list of methods like this